src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+          }
+      }
+    }
+
+  }
+
+  /**
+   * Get reward and next state based on current
+   * state and current action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @param nextState The next state.
+   * @param transitionGoal The goal for transition
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state,
+                const Action& action,
+                State& nextState,
+                const State& transitionGoal)
+  {
+    // Update the number of steps performed.
+    stepsPerformed++;
+
+    // Make a vector to estimate nextstate.
+    arma::vec currentState {state.Row(), state.Column()};
+    arma::vec direction = directions[action.action];
+
+    arma::vec currentNextState = currentState + direction;
+    nextState.Row() = currentNextState[0];
+    nextState.Column() = currentNextState[1];
+
+    // dont move to that position if invalid
+    bool invalid = false;
+
+    if(currentNextState(0)>= maze.n_rows || currentNextState(1)>= maze.n_cols
+        || currentNextState(0) < 0 || currentNextState(1) < 0 || 
+        maze(currentNextState(0),currentNextState(1))==-1 ){
+        invalid = true;
+        nextState.Row() = currentState[0];
+        nextState.Column() = currentState[1];
+    }   
+
+    // Check if the episode has terminated.
+    bool done = IsTerminal(nextState);
+
+    // Do not reward agent if it failed.
+    if (done && maxSteps != 0 && stepsPerformed >= maxSteps)
+      return 0.0;
+    
+    if (done && maze(nextState.Row(),nextState.Column()) == 1 && 
+        nextState.Row() == goal(0) && nextState.Column() == goal(1) )
+      return 1.0;
+
+    return 0.0;
+  }
+
+  /**
+   * Get reward based on current state and current
+   * action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @param transitionGoal The transition goal.
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state, const Action& action,
+                 const State& transitionGoal)
+  {
+    State nextState;
+    return Sample(state, action, nextState, transitionGoal);
+  }
+
+  /**
+   * Initial state (row,column) for agent in maze
+   *
+   * @return Initial state for each episode.
+   */
+  State InitialSample()
+  {
+    stepsPerformed = 0;
+    size_t index = arma::randi(arma::distr_param(0, startingPoints.size() -1));
+    initialState = arma::vec({startingPoints[index][0],startingPoints[index][1]});
+    return State(initialState);
+  }
+
+  /**
+   * Get reward for particular goal
+   *
+   * @param nextState The next state.
+   * @param transitionGoal Transition's goal.
+   * @return Initial state for each episode.
+   */
+  double GetHERReward(const State& nextState,
+                    const State& transitionGoal)
+  {
+    if (maze(nextState.Row(),nextState.Column()) == 0 && 
+         nextState.Row() == transitionGoal.Row() && 
+         nextState.Column() == transitionGoal.Column() )
```

Comment:
see prev comment

Reply:
Its going to give me a vector rather than a bool.

Reply:
Prev comment about making a private function